### PR TITLE
[FIX] prevent unseful writing on products in product_profile

### PR DESCRIPTION
**Pull request by bealdav**
The objective is to fix performance issue.
False changes in product.profile can bring to a lot of write on product.product without this patch.

ping @sebastienbeau @chafique-delli

### CODE REVIEW

**Comment by chafique-delli** [comment]
Thank you @bealdav  for this fix.
 :+1:

File: product_profile/product.py [hunk 1]
```diff
@@ -72,23 +75,47 @@ def _get_types(self):
...
-                keys2remove.append(key)
-        for key in keys2remove:
-            # we remove keys which have no matching in product template
-            vals.pop(key)
-        if vals:
-            products = self.env['product.product'].search(
-                [('profile_id', '=', self.id)])
-            if products:
-                products.write({'profile_id': self.id})
+                    key in get_profile_fields_to_exclude() or
```
Comment (hparfr) [inline product_profile/product.py]: get_profile_fields_to_exclude() should be put outside the loop

File: product_profile/product.py [hunk 2]
```diff
@@ -72,23 +75,47 @@ def _get_types(self):
...
+                    self.check_unuseful_key_in_vals(new_vals, key)):
+                new_vals.pop(key)
+        # super call must be after check_unuseful_key_in_vals() call
+        # because we compare value before and after write
+        res = super(ProductProfile, self).write(vals)
```
Comment (hparfr) [inline product_profile/product.py]: new_vals instead of vals ?

File: product_profile/product.py [hunk 3]
```diff
@@ -72,23 +75,47 @@ def _get_types(self):
...
+                    [('profile_id', '=', rec.id)])
+                if products:
+                    _logger.info(
+                        " >>> %s Products updating after updated '%s' "
+                        "product profile" % (len(products), rec.name))
+                    products.write({'profile_id': rec.id})
         return res
 
     @api.model
+    def check_unuseful_key_in_vals(self, vals, key):
```
Comment (hparfr) [inline product_profile/product.py]: 'useless' instead of 'unuseful' ?

File: product_profile/product.py [hunk 4]
```diff
@@ -72,23 +75,47 @@ def _get_types(self):
...
+            2/ if profile data are in csv file there are processed
+               each time module containing csv is loaded
+            we remove field from vals to minimize impact on products
+        """
+        comparison_value = self[key]
+        if self._fields[key].type == 'many2one':
+            comparison_value = self[key].id
+        elif self._fields[key].type == 'many2many':
+            comparison_value = [(6, False, self[key].ids), ]
+        if vals[key] == comparison_value:
```
Comment (hparfr) [inline product_profile/product.py]: `return vals[key] == comparison_value`

File: product_profile/product.py [hunk 5]
```diff
@@ -72,23 +75,47 @@ def _get_types(self):
...
+        """ If replacing values are the same than in db, we remove them.
+            Use cases:
+            1/ if in edition mode you switch a field
+               from value A to value B and then go back to value A
+               then save form, field is in vals whereas it shouldn't.
+            2/ if profile data are in csv file there are processed
+               each time module containing csv is loaded
+            we remove field from vals to minimize impact on products
+        """
+        comparison_value = self[key]
```
Comment (hparfr) [inline product_profile/product.py:110]: use `else` in the following `if`

**Review by hparfr — COMMENTED** [review]
👍  minor remarks

File: product_profile/product.py (hunk 5 above)
Comment (bealdav) [reply product_profile/product.py:110]: why else instead of elif ?

**Pull request merged**